core/src/main/java/org/elasticsearch/cluster/routing/RoutingNodes.java

```diff
@@ -416,24 +419,162 @@ public ShardRouting initialize(ShardRouting shard, String nodeId, @Nullable Stri
+                // is not added back as unassigned shard
+                removeRelocationSource(targetShard);
+            }
+        }
+
+        // fail actual shard
+        if (failedShard.initializing()) {
+            if (failedShard.relocatingNodeId() == null) {
+                // initializing shard that is not relocation target, just move to unassigned
+                moveToUnassigned(failedShard, unassignedInfo);
+            } else {
+                // The shard is a target of a relocating shard. In that case we only need to remove the target shard and cancel the source
+                // relocation. No shard is left unassigned
+                logger.trace("{} is a relocation target, resolving source to cancel relocation ({})", failedShard,
+                    unassignedInfo.shortSummary());
+                ShardRouting sourceShard = getByAllocationId(failedShard.shardId(),
+                    failedShard.allocationId().getRelocationId());
+                assert sourceShard.isRelocationSourceOf(failedShard);
+                logger.trace("{}, resolved source to [{}]. canceling relocation ... ({})", failedShard.shardId(), sourceShard,
+                    unassignedInfo.shortSummary());
+                cancelRelocation(sourceShard);
+                remove(failedShard);
+            }
+        } else {
+            assert failedShard.active();
+            if (failedShard.primary()) {
+                // promote active replica to primary if active replica exists
+                ShardRouting candidate = activeReplica(failedShard.shardId());
+                if (candidate == null) {
+                    moveToUnassigned(failedShard, unassignedInfo);
+                } else {
+                    movePrimaryToUnassignedAndDemoteToReplica(failedShard, unassignedInfo);
+                    ShardRouting primarySwappedCandidate = promoteAssignedReplicaShardToPrimary(candidate);
+                    if (primarySwappedCandidate.relocatingNodeId() != null) {
+                        // its also relocating, make sure to move the other routing to primary
+                        RoutingNode node = node(primarySwappedCandidate.relocatingNodeId());
+                        if (node != null) {
+                            for (ShardRouting shardRouting : node) {
+                                if (shardRouting.shardId().equals(primarySwappedCandidate.shardId()) && !shardRouting.primary()) {
+                                    promoteAssignedReplicaShardToPrimary(shardRouting);
+                                    break;
+                                }
+                            }
+                        }
+                    }
+                    if (IndexMetaData.isIndexUsingShadowReplicas(indexMetaData.getSettings())) {
+                        reinitShadowPrimary(primarySwappedCandidate);
+                    }
+                }
+            } else {
+                assert failedShard.primary() == false;
```

Comment:
What's the purpose of this assert? Do we expect the `primary()` call to differ between the initial `if` and this?